src/datasource.ts

```diff
@@ -8,34 +8,138 @@ import {
-  async query(options: DataQueryRequest<SumoQuery>): Promise<DataQueryResponse> {
-    const { range, maxDataPoints, targets, scopedVars, interval } = options;
-    const templateSrv = getTemplateSrv();
+  // perform logs/metrics query by Grafana
+  query(options: DataQueryRequest<SumoQuery>): Observable<DataQueryResponse> {
+    const { range , targets } = options;
 
     if (!range) {
       throw new Error('No range has been provided');
     }
+    if(this.logsController$){
+      // stop previous running logs query
+      this.logsController$.next({ type : RunSearchActionType.STOP })
+      this.logsController$ = null;
+    }
+
+    //Check if any logs query present
+    const isLogsQuery = targets.find(query=>query.type === SumoQueryType.Logs)
+
+    if(isLogsQuery){
+      return this.fetchLogsQuery(options);
+    }
+    else{
+      return this.fetchMetricsQuery(options);
+    }
+  }
+
+
+  fetchLogsQuery(options: DataQueryRequest<SumoQuery>) : Observable<DataQueryResponse>{
+
+    const { targets, range } = options;
+    const logsQueryObj  = targets.find(query=>query.type === SumoQueryType.Logs) as SumoQuery
+
+    const startTime = range.from.valueOf();
+    const endTime = range.to.valueOf();
+    const { queryText } = logsQueryObj 
+
+    const searchParams : SearchQueryParams = {
+      types: [SearchQueryType.AGGREGATE],
+      query: queryText || '',
+      startTime : startTime,
+      endTime : endTime,
+      baseUrl : this.baseUrl,
+      basicAuth : this.basicAuth, 
+      paginationInfo : {
+        offset : 0,
+        length : 10000 // fetching first 10000 records only
+      }
+    }
+    const previousCountMap = {
+      messageCount : 0
+    }
+
+    const searchStatusFilter = (status$: Observable<ISearchStatus>) =>
+    status$.pipe(
+      filter((searchStatusResponse: ISearchStatus) => {
+        const isMessageCountUpdated = searchStatusResponse.messageCount && searchStatusResponse.messageCount > previousCountMap.messageCount ? true : false
+        if(isMessageCountUpdated){
+          previousCountMap.messageCount = searchStatusResponse.messageCount as number
+        }
+
+        const isQueryCompleted = searchStatusResponse.state === SearchStatus.DONE_GATHERING_RESULTS
+
+        return (
+          //fetching records when messages count changes and when query gets completed
+          isMessageCountUpdated || isQueryCompleted
+        );
+      }),
+    );
+
+    this.logsController$ = new Subject();
+
+    return searchQueryExecutionService.run(
+      searchParams,
+      this.logsController$,
+      searchStatusFilter
+    ).pipe(switchMap(response=>{
+      const aggregateResponse = response.aggregate
+      if(!aggregateResponse){
+        return of({
+          data  : []
+        }) as  Observable<DataQueryResponse>
+      }
+      const {fields , records} = aggregateResponse
+
+      const dataFrame = new MutableDataFrame({
+        name : 'AggregateResponse', // dataframe name
```

Comment:
I think we also should add

```
meta: {
    preferredVisualisationType: 'logs',
  },
```

Reply:
logs Aggregate query are meant to be visualize on chart. We will add this settings non aggregate results in our next PR